src/allmydata/storage/http_server.py

```diff
@@ -260,21 +269,114 @@ def __init__(self, code: int):
-                    "test": [* {"offset": uint, "size": uint, "specimen": bstr}]
-                    "write": [* {"offset": uint, "data": bstr}]
+                ; TODO Add length limit here, after
+                ; https://github.com/anweiss/cddl/issues/128 is fixed
+                * share_number => {
+                    "test": [*30 {"offset": uint, "size": uint, "specimen": bstr}]
+                    "write": [*30 {"offset": uint, "data": bstr}]
                     "new-length": uint / null
                 }
             }
-            "read-vector": [* {"offset": uint, "size": uint}]
+            "read-vector": [*30 {"offset": uint, "size": uint}]
         }
         share_number = uint
         """
     ),
 }
 
 
-def read_range(request, read_data: Callable[[int, int], bytes]) -> None:
+# Callabale that takes offset and length, returns the data at that range.
+ReadData = Callable[[int, int], bytes]
+
+
+@implementer(IPullProducer)
+@define
+class _ReadAllProducer:
+    """
+    Producer that calls a read function repeatedly to read all the data, and
+    writes to a request.
+    """
+
+    request: Request
+    read_data: ReadData
+    result: Deferred = field(factory=Deferred)
+    start: int = field(default=0)
+
+    @classmethod
+    def produce_to(cls, request: Request, read_data: ReadData) -> Deferred:
+        """
+        Create and register the producer, returning ``Deferred`` that should be
+        returned from a HTTP server endpoint.
+        """
+        producer = cls(request, read_data)
+        request.registerProducer(producer, False)
+        return producer.result
+
+    def resumeProducing(self):
+        data = self.read_data(self.start, 65536)
+        if not data:
+            self.request.unregisterProducer()
+            d = self.result
+            del self.result
+            d.callback(b"")
+            return
+        self.request.write(data)
+        self.start += len(data)
+
+    def pauseProducing(self):
+        pass
+
+    def stopProducing(self):
+        pass
+
+
+@implementer(IPullProducer)
+@define
+class _ReadRangeProducer:
+    """
+    Producer that calls a read function to read a range of data, and writes to
+    a request.
+    """
+
+    request: Request
+    read_data: ReadData
+    result: Deferred
+    start: int
+    remaining: int
+    first_read: bool = field(default=True)
+
+    def resumeProducing(self):
+        to_read = min(self.remaining, 65536)
+        data = self.read_data(self.start, to_read)
+        assert len(data) <= to_read
+        if self.first_read and data:
+            # For empty bodies the content-range header makes no sense since
+            # the end of the range is inclusive.
+            self.request.setHeader(
+                "content-range",
+                ContentRange("bytes", self.start, self.start + len(data)).to_header(),
+            )
+        self.request.write(data)
+
+        if not data or len(data) < to_read:
```

Comment:
This condition might be because `read_data` ran out of bytes to return prematurely.  In this case, the `content-range` header that was written at the beginning will mismatch the actual data returned.  What does HTTP have to say about such a case?

It could also be because that the previous iteration correctly read the last of the data or this iteration is reading the last of it.  In those cases it's perfectly correct to do an orderly completion of the response.

If the bookkeeping at the bottom were done first then `self.remaining == 0`would be a correct condition to use for orderly completion and `len(data) < to_read` would signal the error case.  Maybe it's worth doing something extra in the error case?  The only things I can think of are finishing the request (and maybe this will signal error to the client?  is there another way to signal error?) and logging something.

Reply:
I guess "I returned what's available, and no more" is one possible behavior, and... probably valid.

Reply:
A third option: make sure IStorageServer never reads beyond the end of a share. Which may already be the case but I have distinct memory of some hack used to check existence or something...

Reply:
`test_istorageserver` has some tests that I wrote for reading beyond the edge, so I could just... delete those, and when if I detect usage in client just fix the client? But I swear I had that in for a reason.

Reply:
OK, did some instrumentation, and immutables and mutables do get reads past the end sometimes. So option 3 is no good, and so need to either make HTTP client caller get length first to ensure good range, or just return truncated data.

Reply:
So I think what I will do is, in this branch just return truncated (effectively already implemented) with the wrong Content-Range header in that edge case. And follow up will add two HEAD HTTP endpoints for getting the length of the share, and then HTTP server and client will get logic to not read beyond end of share, and `IStorageServer` layer can use length endpoints to make sure it doesn't request too much.

Reply:
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3907

Reply:
Resolved, but leaving open for context.